Fix run output fd

al8n opened the pull request:
This is a very simple solution. Just add a connection cache based on LRU. This solution can only make sure we will not create multiple conn for one `SocketAddr`.

Zeegomo commented:
I suppose [this](https://github.com/logos-co/nomos-node/issues/313) is problem you're tring to solve.
I'm not sure a cache (especially a time based on) is the best decision. I would expect mixnet nodes to be chosen at random from the upper layer, so there's no correlation between nodes being chosen in the past and nodes being chosen in the future. Also, unless the cache can hold a good chunk of the network nodes I would expect a good number of misses.


I suspect we hit the limit only when trying to send a lot of packets at once, because otherwise sockets should be recycled.
Other possible options:
* batching?
* use other transports
* rate limit creation of new sockets

Zeegomo commented:
> But we are still facing the run out of fd problem if we have too many clients, it seems that each client requires and holds one fd

Yes, but with batching what I meant to say is that if we see that we have to send 10 messages to the same host we can reuse the same connection. Batching also helps with hiding timing information btw

youngjoon-lee commented:
Although I didn't review all changes in this PR yet, my opinion is
- QUIC might be a better solution than TCP. But before we are stable with QUIC, we need to make TCP-based implementation stable.
- For connections between mixnodes, it would be enough to establish/reuse TCP connections with the next mix layer because we will not have the static mixnet topology for now. 
- For connections between mixnodes and mixclients, we may need some solution such as cache or the `connection_limit` option for mixnodes. 
  - I guess the cache solution that you implemented tries to solve this issue.

I wil take a look in detail.

Zeegomo commented:
> For connections between mixnodes, it would be enough to establish/reuse TCP connections with the next mix layer because we will not have the dynamic mixnet topology for now.

It can work provided we can connect to all nodes in the next layer

> For connections between mixnodes and mixclients, we may need some solution such as cache or the connection_limit option for mixnodes.

On the client side, we can probably batch to try reduce the number of connections and then rate limit. Server side we can probably expect a bit more room and/or a deployment with increased OS resources

youngjoon-lee commented:
> It can work provided we can connect to all nodes in the next layer

Yes, the size of mixnet is what I don't expect for now. It's kinda block box. Then I think we can go with the simplest solution.

youngjoon-lee commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,37 @@
+use parking_lot::Mutex;
+use tokio::net::TcpStream;
+
+/// The maximum number of open files allowed for the current process.
+pub static MAX_OPEN_FILES_LIMIT: once_cell::sync::Lazy<Option<usize>> =
+    once_cell::sync::Lazy::new(|| {
+        use rustix::process::{getrlimit, Resource};
+        // minus 1 because we need to leave one file descriptor for the listener
+        getrlimit(Resource::Nofile).maximum.map(|n| n as usize)
+    });
```
In my opinion, it's safer to not have this. Instead, it would be better to force node operators to set the config (instead of `Option`).

youngjoon-lee commented on mixnet/client/src/receiver.rs:
```diff
@@ -1,42 +1,69 @@
 // Receiver accepts TCP connections to receive incoming payloads from the Mixnet.
-pub struct Receiver;
+pub struct Receiver {
+    cache: ConnectionCache,
```
For now, we assume that the mixclient listens messages from a single mixnode (that might be operated by the same guy). Then, I don't think we need cache or something for the `Receiver`.

youngjoon-lee commented on mixnet/client/src/sender.rs:
```diff
@@ -89,15 +98,24 @@ impl<R: Rng> Sender<R> {
-        let mut socket = TcpStream::connect(addr).await?;
-        let body = Body::new_sphinx(packet);
-        body.write(&mut socket).await?;
+        if let Some(stream) = cache.get(&addr) {
+            let mut stream = stream.lock().await;
+            let body = Body::new_sphinx(packet);
+            body.write(&mut *stream).await?;
+        } else {
+            let mut socket = TcpStream::connect(addr).await?;
+            let body = Body::new_sphinx(packet);
+            body.write(&mut socket).await?;
+            cache.insert(addr, Arc::new(Mutex::new(socket)));
+        }
```
I think you already implemented kind of the batching that @Zeegomo mentioned.

p.s.
Even though this mechanism doesn't have buffering (that might reduce the number of calling `write`), I think it's okay. That's the different problem and that's not what Zeegomo suggested, if I understand correctly.

youngjoon-lee replied:
We can probably have `cache.connect_and_get()` for example? so that we can simplify this code block.

Also, it seems there is a possibility that multiple connections to the same host can be established at the line 113, if there are multiple tasks running. This is actually not a big issue since only one will eventually survive, but this can be automatically solved if we have `cache.connect_and_get()`, I guess.

youngjoon-lee commented on mixnet/node/src/lib.rs:
```diff
@@ -21,11 +22,17 @@ use tokio::{
 // A mix node that routes packets in the Mixnet.
 pub struct MixnetNode {
     config: MixnetNodeConfig,
+    cache: ConnectionCache,
```
Actually, the goal of https://github.com/logos-co/nomos-node/issues/313 is not only avoiding the max fd error, but also optimizing the latency of packet forwarding. 

Above, I left a comment about keeping connections for all mixnodes in the next layer, like Nym. But, I think your solution is better when the mixnet is big. In near future, we may need to find the optimal size of mixnet. From Alvaro's suggestion, the size of mixnet will probably affect the privacy. Until then, I'd say let's keep our implementation simple. 
I don't think the cache is very complicated, but I would like to ask for your opinion: cache vs full-static-conns.

Regarding ratelimit or max_conn_limit that we're discussing with @Zeegomo, we need to adopt those concept only for connections between mixclient and mixnode, not for connections between mixnodes. The issue here is that mixnode uses a single port for accepting packets from both mixclients and mixnodes. If we want to adopt ratelimit and max_conn_limit at next PRs, we may need to change this structure somehow.

youngjoon-lee replied:
If we limit the size of mixnet, I think the ratelimit of conn creation in mixclient side is not very necessary.

al8n replied:
Regarding of cache vs full-static-conns, I think if we can make sure we know how many nodes we need to make a connection at the beginning, then I think we can just use lazy initialized full-static-conns.

For packet buffering, I will implement it in the next PR, and keep it the implementation simple (without write ahead log, so we may lose packets when the node crash and buffer still has packets)

youngjoon-lee replied:
Yes. I think buffering is not urgent now. It would be enough to have con reuse mechanism to resolve #313.

youngjoon-lee replied:
> Regarding of cache vs full-static-conns, I think if we can make sure we know how many nodes we need to make a connection at the beginning, then I think we can just use lazy initialized full-static-conns

I agree. I’m fine with any way if it solves the issue simply. If lazy-init is simple to implement for now, I’m okay with it. Bootstrapping all connections at the beginning will probably have its own complexity :) 

At the beginning I thought about just have a hashmap to keep connections, but it’s more or less similar if we set the cache size big :)

al8n commented:
As discussed in Discord with @youngjoon-lee, I replaced LRU with a hash map, and removed the connection cache in the receiver of the mixnet client. If we still need packet buffering and ratelimit, I will implement them in other PRs.

youngjoon-lee commented on mixnet/client/src/receiver.rs:
```diff
@@ -7,21 +7,30 @@ use nym_sphinx::{
-        let Ok(socket) = TcpStream::connect(node_address).await else {
+        let Ok(socket) = TcpStream::connect(self.node_address)
+            .await
+            .map(|s| Arc::new(Mutex::new(s)))
```
maybe we can remove this mutex now?

al8n replied:
In logic, we can remove the mutex, but we still need a mutex to provide internal mutability.

youngjoon-lee replied:
Ah, sorry I don't get. maybe to cover the case when multi threads access the fragment stream?

al8n replied:
Ah, it is because `Body::read(&mut *mu)` requires a mutable reference, and if we remove the Mutex, then we cannot get the mutable reference from `Arc`. We have another `Mutex` on `HashMap`, but we still have to wrap `TcpStream` into an `Arc`, `TcpStream` is not clonable, and we will use this mutable reference in an async closure which requires `'static` lifetime.

youngjoon-lee replied:
Aha. but how about removing both Arc and Mutex from this `receiver.rs`? because we're not gonna use the connection map in this code. I mean reverting this code entirely.

al8n replied:
Aha, yeah, we can totally remove Arc and Mutex, thanks, I forget we are not going to use conn map in this file.

al8n replied:
Aha, yeah, we can totally remove Arc and Mutex, thanks, I forget we are not going to use conn map in this file.

youngjoon-lee commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,26 @@
+
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct ConnectionCache {
+    cache: Arc<Mutex<HashMap<SocketAddr, Arc<tokio::sync::Mutex<TcpStream>>>>>,
+}
```
Now I think we can rename this because this is not a cache anymore :)

As we discussed, I'm fine with both hashmap and LRU at this moment, because they have pros and cons and we don't have any exact plan how to compose the mixnet topology yet :)

youngjoon-lee replied:
In my opinion, one thing we can easily do at next PRs is adding the `max_connections` config to mixnode, so that we can cover the case when there are so many mixclients trying connect to the mixnode. Let's discuss soon.

youngjoon-lee commented on mixnet/node/src/lib.rs:
```diff
@@ -125,31 +138,43 @@ impl MixnetNode {
+        }
 
         let mut socket = TcpStream::connect(addr).await?;
         body.write(&mut socket).await?;
-
+        cache.insert(addr, Arc::new(tokio::sync::Mutex::new(socket)));
```
we can probably use the `get_or_init` that you just added.

al8n replied:
The trade-off about using `get_or_init` here is we have to call `lock`.

```rust
let socket: Arc<Mutex<TcpStream>> = cache.get_or_init(addr).await;
let mut mu = socket.lock().await;
body.write(&mut socket).await?;
```

youngjoon-lee replied:
Yes, we have tradeoff. If we don't want to allow the line 175 to be executed concurrently for the same host, we should use `get_or_init`. In the extreme case, this will cause the 'ran out of fd` error.

youngjoon-lee replied:
As long as we decide to use the map with mutex, it means we're gonna take the overhead of lock contention whenever we handle connections. If so, it's better to have the consistent behaviour everywhere, imho.

youngjoon-lee commented on mixnet/node/src/lib.rs:
```diff
@@ -78,28 +82,36 @@ impl MixnetNode {
         match body {
-            Body::SphinxPacket(packet) => Self::handle_sphinx_packet(private_key, packet).await,
+            Body::SphinxPacket(packet) => {
+                cache.insert(
+                    socket.peer_addr().unwrap(),
+                    Arc::new(tokio::sync::Mutex::new(socket)),
+                );
+                Self::handle_sphinx_packet(cache, private_key, packet).await
```
Ah, do we also need to add inbound connections to the map? I thought about keeping only outbound connections, but let me think.

youngjoon-lee replied:
I think, even if we put inbound connections to the map, we won't use them again because the mixnet topology is static now (In case when we make the mixnet topology dynamic in the future, we will probably clear the map). Also, mixnodes won't send something to mixclient actively.

If we don't store inbound connections to the map, the connections are dropped and clients cannot reuse those connections anymore? Then, it makes sense.

al8n replied:
Yeah, if we drop on the server side, the client side is no longer valid.

al8n replied:
> the server-side may need to read these inbound connections periodically. The simplest solution would be have an infinite loop(task) for each inbound connection instead of putting them into the map. Hm, it seems this topic is more complicated than I expected. Let me run some tests and think about it again.

I totally agree, in this case, we need to periodically read inbound connections, once we accept a conn, we need to spawn a new thread to handle the logic.

youngjoon-lee replied:
Yes. In my opinion, it's better to make the task continuously read the socket (then we don't need to store the socket in the map). From your implementation, it seems the connection is stored in the map and never be read anymore.

youngjoon-lee replied:
For example, like 501b263 (I didn't make tests work).

al8n replied:
Yeah, makes sense. We do not need to store conn in the map anymore.

youngjoon-lee commented on mixnet/node/src/lib.rs:
```diff
@@ -81,18 +81,38 @@ impl MixnetNode {
-    ) -> Result<(), Box<dyn Error>> {
-        let body = Body::read(&mut socket).await?;
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        tokio::pin!(socket);
+        loop {
+            tokio::select! {
+                // TODO: add a random delay?
+                _ = tokio::time::sleep(std::time::Duration::from_millis(100)) => {}
+                body = Body::read(&mut socket) => {
+                    match body {
+                        Ok(body) => if let Err(e) = Self::handle_body(body, &cache, &private_key, &client_tx).await {
+                            tracing::error!("failed to handle body: {e}");
+                        },
+                        Err(e) => {
+                            tracing::error!("failed to read body: {e}");
+                        }
+                    }
+                }
+            }
+        }
```
We may need to give up reading this socket at the line 100, instead of continuing the loop? 

Also, it seems we don't need the sleep. Please let me know if I'm missing something.

youngjoon-lee commented on mixnet/node/src/lib.rs:
```diff
@@ -77,38 +81,66 @@ impl MixnetNode {
-    ) -> Result<(), Box<dyn Error>> {
-        let body = Body::read(&mut socket).await?;
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        tokio::pin!(socket);
+        loop {
+            tokio::select! {
+                // TODO: add a random delay?
+                _ = tokio::time::sleep(std::time::Duration::from_millis(100)) => {}
+                body = Body::read(&mut socket) => {
+                    match body {
+                        Ok(body) => if let Err(e) = Self::handle_body(body, &cache, &private_key, &client_tx).await {
+                            tracing::error!("failed to handle body: {e}");
+                        },
+                        Err(e) => {
+                            tracing::error!("failed to read body: {e}");
+                        }
+                    }
+                }
+            }
+        }
```
We may need to give up reading this socket at the line 100, instead of continuing the loop?

Also, it seems we don't need the sleep. Please let me know if I'm missing something.

al8n replied:
Yeah, after rethinking, we do not need the sleep.

youngjoon-lee commented on mixnet/client/src/config.rs:
```diff
@@ -10,6 +10,7 @@ use crate::{receiver::Receiver, MessageStream, MixnetClientError};
 pub struct MixnetClientConfig {
     pub mode: MixnetClientMode,
     pub topology: MixnetTopology,
+    pub connection_cache_size: usize,
```
We can rename this as well. (also in `MixnetNodeConfig`).

youngjoon-lee commented on mixnet/client/src/lib.rs:
```diff
@@ -22,17 +23,22 @@ pub type MessageStream = BoxStream<'static, Result<Vec<u8>, MixnetClientError>>;
 
 impl<R: Rng> MixnetClient<R> {
     pub fn new(config: MixnetClientConfig, rng: R) -> Self {
+        let cache = ConnectionPool::new(config.connection_cache_size);
```
We can also rename this variable.

youngjoon-lee commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,42 @@
+        self.pool.lock().get(addr).cloned()
+    }
+
+    #[allow(clippy::await_holding_lock)]
+    pub async fn get_or_init(
+        &self,
+        addr: &SocketAddr,
+    ) -> std::io::Result<Arc<tokio::sync::Mutex<TcpStream>>> {
+        let mut pool = self.pool.lock();
+        match pool.get(addr).cloned() {
+            Some(tcp) => Ok(tcp),
+            None => {
+                let tcp = Arc::new(tokio::sync::Mutex::new(TcpStream::connect(addr).await?));
```
Ah ah sorry, I didn't notice that we have `connect().await` while acquiring lock.

After rethinking, it might not be very important to prevent temporary multiple connections for the same host because it will rarely happen probably. It's not easy to predict how critial this would be. Let's see and improve this gradually. Please feel free to correct me if I'm missing something.

youngjoon-lee reviewed (approved):
Thank you! I think this is a good starting point. From this, we can improve the connection management gradually, such as reconnect, retry, max_conn_limit, idle_timeout, buffering, and so on. (I'm not saying we need all of them. Just writing down what comes to my mind now).
I left some topics in https://github.com/logos-co/nomos-node/issues/322, and we can also get hints from [Nym implementation](https://github.com/nymtech/nym/blob/3748ab77a132143d5fd1cd75dd06334d33294815/common/client-libs/mixnet-client/src/client.rs#L197-L197) (Nym has most of them above, as I remember). Or even, we can consider QUIC.

For now, I think we can go with this simple strategy which reduces the `ran out of fds` error and TCP connect handshakes.

Zeegomo commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,42 @@
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct ConnectionPool {
+    pool: Arc<Mutex<HashMap<SocketAddr, Arc<tokio::sync::Mutex<TcpStream>>>>>,
```
how's keepalive handled by the pool?

al8n replied:
Yeah, currently we do not check if the conn is still living. As @youngjoon-lee mentioned, the implementation is not fail-safe. I will implement dead connection checks and cleanup in the next PR.

youngjoon-lee replied:
Although this can be done in the next PR, I guess the easiest way might be the "on-demand reconnection", instead of checking the status of connections in the pool periodically.

Zeegomo reviewed (changes requested):
Apart from the comments, I'm still having doubts about the solution as a whole. 
Given that routes through the mixnet are supposed to be random or at least we can't expect specific patterns, keeping a subset of connections might not be very efficient (exact numbers depend on subset size ofc).
If we can guarantee the pool to be as big as necessary (e.g. the next layer) then this is good, but otherwise I would like to see measurements to see whether this helps or not and then some documentation on our findings.
Actually, I do not see any limit on the size of the connection pool, so it grows as big as necessary atm, is that your intention or you forgot to add it (there's some mention of it in the doc but that parameter does not actually do anything)?

al8n commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,42 @@
+#[derive(Clone)]
+#[repr(transparent)]
+pub struct ConnectionPool {
+    pool: Arc<Mutex<HashMap<SocketAddr, Arc<tokio::sync::Mutex<TcpStream>>>>>,
```
`Body::read` or `Body::write` requires a mutable reference, and we also use async closure, which requires a `'static` lifetime, and `TcpStream` is not clonable. Although we know that there is no data race in `TcpStream`. There seems to be no safe way to implement internal mutability without `Mutex`.

Zeegomo replied:
I think the external one should be sufficient to return a `&mut TcpStream`? Async closures could just get a reference to the external struct

al8n replied:
If we return a `&mut TcpStream`, then we must hold the `MutexGuard` of the whole pool, it seems that we are not parallel anymore.

al8n commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,42 @@
+}
+
+impl ConnectionPool {
+    pub fn new(size: usize) -> Self {
```
Sorry for the misleading. I will update the doc, because currently, we assume the mix node in each layer will not be too large. The size here just helps initialize the `HashMap` of the pool, and no eviction policy is applied.

youngjoon-lee replied:
> Apart from the comments, I'm still having doubts about the solution as a whole.
Given that routes through the mixnet are supposed to be random or at least we can't expect specific patterns, keeping a subset of connections might not be very efficient (exact numbers depend on subset size ofc).
If we can guarantee the pool to be as big as necessary (e.g. the next layer) then this is good, but otherwise I would like to see measurements to see whether this helps or not and then some documentation on our findings.
Actually, I do not see any limit on the size of the connection pool, so it grows as big as necessary atm, is that your intention or you forgot to add it (there's some mention of it in the doc but that parameter does not actually do anything)?

I think we can discuss about this topic here. 

Yeah, since the current implementation is the simplest approach, we don't have eviction. Though, I thought about boostrapping all connections with the next mix layer at the beginning. Instead, now we do lazy-init. They have pros and cons, but lazy-init might be easier to implement for now. 
For now, IMO, it's enough to set the max size of hashmap to the size of the next mix layer. If it hits the max, we could panic or return an error. That won't happen unless the sender picked the wrong route.

youngjoon-lee replied:
We can simply run this test: https://github.com/logos-co/nomos-node/pull/326/files#diff-de86edb550dad603a64ce6e7b47a7e62f68a2b91ddc8c6e96dd858cb4a47c69bR55, to see if this PR solves the `ran out of fds` error. Plus, we need to check if this PR improves the overall latency.

Zeegomo replied:
if this solution is a lazy-init then it's fine (although we have some 2 locks to access a connection).
Also atm we don't have any limit on the max size of the hashmap, but also I do not seed a need for it

al8n commented:
> Apart from the comments, I'm still having doubts about the solution as a whole. Given that routes through the mixnet are supposed to be random or at least we can't expect specific patterns, keeping a subset of connections might not be very efficient (exact numbers depend on subset size ofc). If we can guarantee the pool to be as big as necessary (e.g. the next layer) then this is good, but otherwise I would like to see measurements to see whether this helps or not and then some documentation on our findings. Actually, I do not see any limit on the size of the connection pool, so it grows as big as necessary atm, is that your intention or you forgot to add it (there's some mention of it in the doc but that parameter does not actually do anything)?

Yes, the current solution is based on the assumption that the mix node in each layer will not be too large and the number of conns will always fit with the pool.

al8n commented on mixnet/util/src/lib.rs:
```diff
@@ -0,0 +1,42 @@
+use tokio::net::TcpStream;
+
+#[derive(Clone)]
+#[repr(transparent)]
```
Ah, I get used to mark a single field struct with `#[repr(transparent)]`. But I do not have a deep understanding of the pros and cons.

Zeegomo replied:
It tells the compiler that you want the in-memory representation of the struct to be exactly the same as the only field. It's only useful if you need to inspect raw representation or do FFI and it could actually keep the compiler from doing optimizations (or you from adding a new field without having to remove it). 

Unless we have a clear reason to have this, let the compiler figure out what's best

al8n replied:
Thanks! 👍
